### PR TITLE
Suggest using `as_ref` on `*const T`

### LINKED ISSUE #21596: Misleading message when attempting to call methods on a raw pointer
When trying to call a method on a raw pointer Rust compiler says there's no such method.

It would be better to say that there is such method, but the pointer has to be explicitly dereferenced first.

``` rust
let x = 8u8;
let y = &x;
let z: *const u8 = &x;
println!("{} {}", 
    y.to_string(), // OK
    z.to_string()  // Not OK
);
```

> error: type `*const u8` does not implement any method in scope named `to_string`

This is _technically correct_, but unexpected and help hint for this error suggests wrong solution. 

I've read this message as "yada yada _no such method_", so I went to check whether I've spelled the method correctly, whether I've put the method in the appropriate `impl` block, etc. — all completely irrelevant to this error.

I'd find something like that more helpful:

> error: method `to_string` requires type `u8`, but `*const u8` cannot be automatically dereferenced. hint: use `as_ref()`

### PR DESCRIPTION
**Pull request by estebank**
Fix #21596.

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -324,6 +324,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 );
                             }
                         }
+                        if let ty::RawPtr(_) = &actual.sty {
+                            err.note("you might want to use `<*const T>::as_ref()` to get a \
```
Comment (Centril) [inline src/librustc_typeck/check/method/suggest.rs]:
```suggestion
                            err.note("try using `<*const T>::as_ref()` to get a \
```
A bit more direct & succinct =)

Comment (estebank) [reply src/librustc_typeck/check/method/suggest.rs]: done

**Comment by RalfJung** [comment]
Might be a good place to mention that *this is UB* if the pointer is dangling, unaligned or does not point to initialized memory -- even if the result reference is never used?

**Comment by estebank** [comment]
@ralfjung sounds reasonable. Can you file a follow up ticket?

**Comment by RalfJung** [comment]
I was thinking maybe adding that to the `as_ref` docs makes more sense? Or both?

I have adding it to the docs on my list.

**Comment by estebank** [comment]
I think minimizing the _requirement_ of having the docs open is a good thing. Basically anything I would tell someone when pair programming should be in the diagnostic itself, and this would count. I feel the diagnostics are part of the documentation, and the clarification should certainly be in the proper docs as well.

**Comment by RalfJung** [comment]
Issue opened at https://github.com/rust-lang/rust/issues/61786.

**Pull request merged**